src/libcore/intrinsics.rs

```diff
@@ -1335,6 +1337,29 @@ mod real_intrinsics {
   }
 }
 
+/// Checks whether `ptr` is properly aligned with respect to
+/// `align_of::<T>()`.
+pub(crate) fn is_aligned_and_not_null<T>(ptr: *const T) -> bool {
+    return !ptr.is_null() && ptr as usize % mem::align_of::<T>() == 0;
+}
+
+/// Checks whether the regions of memory starting at `src` and `dst` of size
+/// `count * size_of::<T>()` overlap.
+fn overlaps<T>(src: *const T, dst: *const T, count: usize) -> bool {
+    use crate::cmp::Ordering;
+    let src_usize = src as usize;
+    let dst_usize = dst as usize;
+    let size = mem::size_of::<T>() * count;
+    match src_usize.cmp(&dst_usize) {
```

Comment:
Hey, you should probably return
abs(src_usize - dst_usize) < size
It's less convoluted and should cover all cases of src and dst overlaping.

Reply:
You're right. The absolute diff of 2 usize is a bit trickier though, so I used a simple if.